test/integration/ws/w_socket_common.dart

```diff
@@ -0,0 +1,293 @@
+        socket = await WSocket.connect(pingUri);
+        WSHelper helper = new WSHelper(socket);
+
+        socket.add('ping2');
+        await helper.messagesReceived(2);
+
+        expect(helper.messages, unorderedEquals(['pong', 'pong']));
+      });
+
+      test('should not allow multiple listeners by default', () async {
+        socket = await WSocket.connect(echoUri);
+        socket.listen((_) {});
+        expect(() {
+          socket.listen((_) {});
+        }, throwsStateError);
+      });
+
+      test('should not miss messages if a listener is registered late',
+          () async {
+        socket = await WSocket.connect(pingUri);
+        socket.add('ping3');
+        await new Future.delayed(new Duration(milliseconds: 200));
+        expect(socket.toList(), completion(equals(['pong', 'pong', 'pong'])));
+        socket.close();
+      });
+
+      test('should call onDone() when socket closes', () async {
+        socket = await WSocket.connect(echoUri);
+
+        Completer c = new Completer();
+        socket.listen((_) {}, onDone: () {
+          c.complete();
+        });
+
+        socket.close();
+        await c.future;
+      });
+
+      test(
+          'should have the close code and reason available in onDone() callback',
+          () async {
+        socket = await WSocket.connect(echoUri);
+
+        Completer c = new Completer();
+        socket.listen((_) {}, onDone: () {
+          expect(socket.closeCode, equals(4001));
+
+          // TODO: Dart's WebSocket server did not previously set the closeReason
+          // See: https://github.com/dart-lang/sdk/issues/23964
+          // expect(socket.closeReason, equals('Closed.'));
+
+          c.complete();
+        });
+
+        socket.add('echo');
+
+        new Timer(new Duration(seconds: 1), () {
+          socket.close(4001, 'oops');
+        });
+
+        await c.future;
+      });
+
+      test('should work as a broadcast stream', () async {
+        socket = await WSocket.connect(pingUri);
+        Stream stream = socket.asBroadcastStream();
+
+        Completer c1 = new Completer();
+        Completer c2 = new Completer();
+
+        stream.listen((_) {
+          c1.complete();
+        });
+        stream.listen((_) {
+          c2.complete();
+        });
+
+        socket.add('ping');
+
+        await c1.future;
+        await c2.future;
+      });
+    });
+
+    group('closing', () {
+      test('should have the close code and reason available after closing',
+          () async {
+        socket = await WSocket.connect(echoUri);
+        await socket.close(4001, 'Closed.');
+        expect(socket.closeCode, equals(4001));
+
+        // TODO: Dart's WebSocket server did not previously set the closeReason
+        // See: https://github.com/dart-lang/sdk/issues/23964
+        // expect(socket.closeReason, equals('Closed.'));
+      });
+
+      test(
+          'should close and properly drain stream even if no listeners were registered',
```

Comment:
When a websocket is created, we create a custom stream that is mapped to the underlying `WebSocket` instance. The intermediary subscription is immediately paused and is only resumed when the first listener subscribes. If no listeners are registered, this never happens, and the subscription would remain in this paused state if no other action was taken. But, we listen for the close event and drain that intermediary subscription so that those resources don't linger. This is tested by waiting for the `.close()` future to resolve, which only happens when that intermediary subscription has finished.